### ISSUE TITLE
Changing recipe in init file doesn't load new recipe for a package

### ISSUE THREAD

**Issue by vyp** [raxod502/straight.el#166]
I'm using straight.el (master branch) with use-package and have the following in my `init.el`:

``` elisp
(setq use-package-always-ensure t)
(use-package general :demand t)
```

Which works, general gets installed. But now I want to use a different branch for that package, so I change the second line to:

``` elisp
(use-package general :demand t
  :recipe (:host github :repo "noctuid/general.el" :branch "buttercup"))
```

Now when I quit and relaunch emacs, I don't think I see any new effect, general is still on the default master branch, not buttercup. I may have missed it in the docs, but is this expected behaviour?

**Comment by dieggsy** [comment]
I think I've noticed this as well - if you call `straight-pull-package` on `general`, it will ask to update the remote. I do think this prompt should probably happen on startup, since the idea is that the init.el dictates the "state" of your emacs config.

**Comment by vyp** [comment]
Yeah exactly. In fact, I don't even think there should be a prompt and it should 'just work'. For example, if I didn't have general installed previously, and just put the second code snippet in my init.el, then general gets installed and is on the buttercup branch. So I expect the same result if I have the same thing in my init.el. But to have a prompt would mean different behaviour based on some previous state (i.e. the fact that I installed general beforehand with default recipe), and so the init file is no longer the sole source of truth.

**Comment by raxod502** [comment]
This is expected behavior. Otherwise, it would not be possible to switch branches without updating the recipe in your init-file. Furthermore, it would be impossible to guarantee a fast startup if it were necessary to invoke `git(1)` once for every repository being loaded.

The idea is that your init-file specifies a single source of truth, which may have _arbitrary local modifications applied on top of it_. To revert those local modifications, use `M-x straight-normalize-all`. However, it doesn't make sense to revert them automatically because (1) that would lead to data loss, and (2) it would not be possible, performance-wise.

See also https://github.com/raxod502/straight.el/issues/103#issuecomment-325833346 for more discussion.

In your case @vyp, you would want to `M-x straight-normalize-package RET general RET` in order to switch branches.

I'm open to suggestions on ways in which the user experience or documentation could be improved in this area.

**Comment by vyp** [comment]
Yeah so after I made my last comment, I was thinking that requiring such a check at every startup would be way too costly in terms of performance, but then if I for example enter a new package in my init.el, then straight does install it at the next startup, so what's going on here? How does straight do this without being slow, is it a simple case of checking if the local repo exists? (I know can just look at the source code, but thought I might ask as well.)

re: local modifications, I thought about this, and I think the ability to load the local state of the directory is nice and like the way you designed it to be used. For example, even if it's not completely *deterministic* from the init file, it can be made deterministic pretty easily, and plus this way there's no overhead to just experimenting with quick (or long) arbitrary changes, sort of the way emacs is supposed to be.

> I'm open to suggestions on ways in which the user experience or documentation could be improved in this area.

So for me it wasn't clear that I had to use `normalize` at all, since from the readme it seemed as though it would only be useful if I make local modifications. I guess my confusion comes from the fact that entering a new package into the init.el will install it without needing to do anything else at the next startup, but changing a recipe for an already installed package doesn't do anything. That's not to say I don't want straight to install new packages at the next startup, I like the way it works now, but I was just trying to explain how I came to misunderstand the expected behaviour.

I'm not sure how to address that in the documentation though, but does that help? Maybe, a comment somewhere saying `normalize` is also useful to realize some changes to the init file, such as modifying a recipe..? (I'm happy to close this issue too if you'd like.)

**Comment by raxod502** [comment]
> is it a simple case of checking if the local repo exists?

[Yes.](https://github.com/raxod502/straight.el/blob/4df854429a3ed002b40b183bbc0c89406b9fe743/straight.el#L1354)

> this way there's no overhead to just experimenting with quick (or long) arbitrary changes, sort of the way emacs is supposed to be.

In other words, the whole reason I made this package manager.

> So for me it wasn't clear that I had to use `normalize` at all, since from the readme it seemed as though it would only be useful if I make local modifications.

Agreed, this is not made clear in the README.

> entering a new package into the init.el will install it without needing to do anything else at the next startup, but changing a recipe for an already installed package doesn't do anything

It's because there's no way to distinguish a recipe change from a local modification, or some combination of both.

> Maybe, a comment somewhere saying `normalize` is also useful to realize some changes to the init file, such as modifying a recipe..?

Good idea. This should be made obvious, since it only makes immediate sense from the developer's perspective (i.e. mine), and not the user's.

> (I'm happy to close this issue too if you'd like.)

Let's leave it open to track the documentation improvements.

**Comment by hrehfeld** [comment]
Is this actually documented somewhere? I looked for about 10 minutes how to switch from original package to my fork, and ended up just adding the remote manually.

**Comment by raxod502** [comment]
Adding `:fork t`, reloading your init-file, and then normalizing should cause the remote to be added. Doing `straight-fetch-package` should fetch from your forked remote. You would also have to `straight-merge-package` to actually update HEAD to the later commit if you have more stuff on the main branch in your fork.

In terms of documentation, I think the information is technically available in the README, but not all in one place like is necessary to understand it holistically. Sorry, I feel bad for creating such a complex system.

**Comment by raxod502** [comment]
What you're suggesting is very plausible without the need for tradeoffs around data loss, because `straight.el` supports, by design, interactive reconciliation of local changes with what is specified in the recipe and lockfile. It would be similar to running `M-x straight-normalize-package` when the last known recipe from the build cache is shown to be outdated. The build cache is not guaranteed to be persistent, so some additional work would be needed to handle all cases correctly, but this is very feasible.
